### PR TITLE
fix readdirplus when filler is called with zero offset

### LINKED ISSUE #235: How can I tell that readdirplus is working?
From what I understand, readdirplus allows calls to `readdir()` to also include stat information which will be cached.  Thus, one `ls -l` command will not require context switching into libfuse userland for each and every file's stat information.  Given this, I would expect there to be fewer `LOOKUP`s generated with `ls -l` and `find` when a libfuse program is running with `-d`, fewer calls to `getattr()`, and improved `find` times.  However, I'm not seeing a difference.  It is not clear to me if I am misunderstanding readdirplus, misusing fuse, or there's a bug.

Using:

- libfuse 3.2.1, compiled from source
- Linux 4.9.0 (from Debian Stretch)

Using `passthrough_fh.c` as a reference, I run:

	mkdir -p /example
	touch /example/{foo,bar,baz}
	./passthrough_fh -d /dev/shm/passothrough

then in another terminal:

	ls -l ./test/example/

passthrough_fh prints:

	FUSE library version: 3.2.1
	nullpath_ok: 0
	unique: 1, opcode: INIT (26), nodeid: 0, insize: 56, pid: 0
	INIT: 7.26
	flags=0x001ffffb
	max_readahead=0x00020000
	   INIT: 7.26
	   flags=0x0000f439
	   max_readahead=0x00020000
	   max_write=0x00020000
	   max_background=0
	   congestion_threshold=0
	   time_gran=1
	   unique: 1, success, outsize: 80
	unique: 2, opcode: LOOKUP (1), nodeid: 1, insize: 48, pid: 20375
	LOOKUP /example
	getattr[NULL] /example
	   NODEID: 2
	   unique: 2, success, outsize: 144
	unique: 3, opcode: GETATTR (3), nodeid: 2, insize: 56, pid: 20375
	getattr[NULL] /example
	   unique: 3, success, outsize: 120
	unique: 4, opcode: LOOKUP (1), nodeid: 1, insize: 48, pid: 20375
	LOOKUP /example
	getattr[NULL] /example
	   NODEID: 2
	   unique: 4, success, outsize: 144
	unique: 5, opcode: GETXATTR (22), nodeid: 2, insize: 65, pid: 20375
	   unique: 5, error: -38 (Function not implemented), outsize: 16
	unique: 6, opcode: LOOKUP (1), nodeid: 1, insize: 48, pid: 20375
	LOOKUP /example
	getattr[NULL] /example
	   NODEID: 2
	   unique: 6, success, outsize: 144
	unique: 7, opcode: LOOKUP (1), nodeid: 1, insize: 48, pid: 20375
	LOOKUP /example
	getattr[NULL] /example
	   NODEID: 2
	   unique: 7, success, outsize: 144
	unique: 8, opcode: OPENDIR (27), nodeid: 2, insize: 48, pid: 20375
	opendir flags: 0x18800 /example
	   opendir[39812704] flags: 0x18800 /example
	   unique: 8, success, outsize: 32
	unique: 9, opcode: GETATTR (3), nodeid: 2, insize: 56, pid: 20375
	getattr[NULL] /example
	   unique: 9, success, outsize: 120
	unique: 10, opcode: READDIRPLUS (44), nodeid: 2, insize: 80, pid: 20375
	readdirplus[39812704] from 0
	   unique: 10, success, outsize: 816
	unique: 11, opcode: LOOKUP (1), nodeid: 1, insize: 48, pid: 20375
	LOOKUP /example
	getattr[NULL] /example
	   NODEID: 2
	   unique: 11, success, outsize: 144
	unique: 12, opcode: LOOKUP (1), nodeid: 2, insize: 44, pid: 20375
	LOOKUP /example/bar
	getattr[NULL] /example/bar
	   NODEID: 3
	   unique: 12, success, outsize: 144
	unique: 13, opcode: GETATTR (3), nodeid: 3, insize: 56, pid: 20375
	getattr[NULL] /example/bar
	   unique: 13, success, outsize: 120
	unique: 14, opcode: LOOKUP (1), nodeid: 1, insize: 48, pid: 20375
	LOOKUP /example
	getattr[NULL] /example
	   NODEID: 2
	   unique: 14, success, outsize: 144
	unique: 15, opcode: LOOKUP (1), nodeid: 2, insize: 44, pid: 20375
	LOOKUP /example/baz
	getattr[NULL] /example/baz
	   NODEID: 4
	   unique: 15, success, outsize: 144
	unique: 16, opcode: GETATTR (3), nodeid: 4, insize: 56, pid: 20375
	getattr[NULL] /example/baz
	   unique: 16, success, outsize: 120
	unique: 17, opcode: LOOKUP (1), nodeid: 1, insize: 48, pid: 20375
	LOOKUP /example
	getattr[NULL] /example
	   NODEID: 2
	   unique: 17, success, outsize: 144
	unique: 18, opcode: LOOKUP (1), nodeid: 2, insize: 44, pid: 20375
	LOOKUP /example/foo
	getattr[NULL] /example/foo
	   NODEID: 5
	   unique: 18, success, outsize: 144
	unique: 19, opcode: GETATTR (3), nodeid: 5, insize: 56, pid: 20375
	getattr[NULL] /example/foo
	   unique: 19, success, outsize: 120
	unique: 20, opcode: READDIRPLUS (44), nodeid: 2, insize: 80, pid: 20375
	readdirplus[39812704] from 9223372036854775807
	   unique: 20, success, outsize: 16
	unique: 21, opcode: RELEASEDIR (29), nodeid: 2, insize: 64, pid: 0
	releasedir[39812704] flags: 0x0
	   unique: 21, success, outsize: 16

I would have expected, were readdirplus working, that the `LOOKUP` lines corresponding to `foo`, `bar`, and `baz` would be absent.

I modified `passthrough_fh` to include an indication if it is trying to use readdirplus:

	  static int xmp_readdir(const char *path, void *buf, fuse_fill_dir_t filler,
	  		       off_t offset, struct fuse_file_info *fi,
	  		       enum fuse_readdir_flags flags)
	  {
	+ 	printf("using readdirplus: %d\n", flags & FUSE_READDIR_PLUS);
	  	struct xmp_dirp *d = get_dirp(fi);

and to indicate if it's *actually* calling the underlying getattr()

	  static int xmp_getattr(const char *path, struct stat *stbuf,
	  			struct fuse_file_info *fi)
	  {
	+ 	printf("calling getattr for %s\n", path);
	  	int res;

Then rebuilt and ran:

	./passthrough_fh -f /dev/shm/passothrough

then ran the `ls -l` example above. The modified `passthrough_fh` printed:

	calling getattr for /example
	calling getattr for /example
	calling getattr for /example
	calling getattr for /example
	calling getattr for /example
	calling getattr for /example
	using readdirplus: 1
	calling getattr for /example
	calling getattr for /example/bar
	calling getattr for /example/bar
	calling getattr for /example
	calling getattr for /example/baz
	calling getattr for /example/baz
	calling getattr for /example
	calling getattr for /example/foo
	calling getattr for /example/foo
	using readdirplus: 1

If the kernel did not want to use readdirplus, I would have expected `flags & FUSE_READDIR_PLUS` to be `0`, and thus it does seem like readdirplus is being requested.

Were readdirplus working, I would have expected the `getattr` calls for `foo`, `bar`, and `baz` to be absent and the preceding `readdir()` would have populated those values.

Past releases of libfuse apparently supported a `-o readirplus=(yes|no|auto)` that may have been useful for testing here, but 3.2.1 does not appear to do so.

I have also tried modifying `xmp_readdir` in various ways to try and disable readirplus (e.g. placing `flags=0;` at the very beginning), performing quick and dirty benchmarks with `find` with and without the modification.  I saw no meaningful difference, although my testing methodology may have been flawed.

What am I missing here?

### PR DESCRIPTION
**Pull request by farlongsignal**
fixes #235 
In fill_dir_plus(), there's a lookup for caching dirent attributes. However, when offset = 0 the cache metadata from the lookup is lost as only the entry attributes are passed when added to the list. Kernel probably ignores the attributes since .ino = 0.
With the following PR i can finally witness file attributes being cached and getattr() not being called when running `ls -l`.

### CODE REVIEW

File: lib/fuse.c [hunk 1]
```diff
@@ -3673,6 +3676,18 @@ static int readdir_fill_from_list(fuse_req_t req, struct fuse_dh *dh,
 				.ino = 0,
 				.attr = de->stat,
 			};
+
+			if (flags & FUSE_READDIR_PLUS) {
```
Comment (bsbernd) [inline lib/fuse.c:3680]: I need to go over all details of this patch and readdir-plus code, but here you are already in the condition `if (flags & FUSE_READDIR_PLUS)` - no need to check another times.

**Review by bsbernd — CHANGES_REQUESTED** [review]
Please also add a more explaining commit message (and ideally a real commit author name).

**Comment by bsbernd** [comment]
Hmm, I see the commit message here in `Conversation`, but it seems to be missing for the commit itself?

File: lib/fuse.c (hunk 1 above)
Comment (farlongsignal) [reply lib/fuse.c]: thanks, fixed.
Please take your time, i'm obviously new to libfuse codebase so i might be misunderstanding some things.

**Comment by farlongsignal** [comment]
> Hmm, I see the commit message here in `Conversation`, but it seems to be missing for the commit itself?

included commit message

**Comment by bsbernd** [comment]
@farlongsignal  I just tested your patch with `./passthrough_fh -d -oentry_timeout=60 /tmp/mnt/` and somehow I don't see a difference.

Patched:

```bschubert2@imesrv6 example>./passthrough_fh -d -oentry_timeout=60 /tmp/mnt/
FUSE library version: 3.16.2
nullpath_ok: 0
unique: 2, opcode: INIT (26), nodeid: 0, insize: 104, pid: 0
INIT: 7.38
flags=0x73fffffb
max_readahead=0x00020000
   INIT: 7.39
   flags=0x4040f439
   max_readahead=0x00020000
   max_write=0x00100000
   max_background=0
   congestion_threshold=0
   time_gran=1
   unique: 2, success, outsize: 80
unique: 4, opcode: GETATTR (3), nodeid: 1, insize: 56, pid: 3837234
getattr[NULL] /
   unique: 4, success, outsize: 120
unique: 6, opcode: GETXATTR (22), nodeid: 1, insize: 65, pid: 3837234
getxattr / security.selinux 255
   unique: 6, error: -61 (No data available), outsize: 16
unique: 8, opcode: GETXATTR (22), nodeid: 1, insize: 72, pid: 3837234
getxattr / system.posix_acl_access 4096
   unique: 8, error: -61 (No data available), outsize: 16
unique: 10, opcode: GETXATTR (22), nodeid: 1, insize: 73, pid: 3837234
getxattr / system.posix_acl_default 4096
   unique: 10, error: -61 (No data available), outsize: 16
unique: 12, opcode: OPENDIR (27), nodeid: 1, insize: 48, pid: 3837234
opendir flags: 0x18800 /
   opendir[140180484788512] flags: 0x18800 /
   unique: 12, success, outsize: 32
unique: 14, opcode: GETATTR (3), nodeid: 1, insize: 56, pid: 3837234
getattr[NULL] /
   unique: 14, success, outsize: 120
unique: 16, opcode: READDIRPLUS (44), nodeid: 1, insize: 80, pid: 3837234
readdirplus[140180484788512] from 0
   unique: 16, success, outsize: 4016
unique: 18, opcode: LOOKUP (1), nodeid: 1, insize: 44, pid: 3837234
LOOKUP /var
getattr[NULL] /var
   NODEID: 2
   unique: 18, success, outsize: 144
unique: 20, opcode: GETATTR (3), nodeid: 2, insize: 56, pid: 3837234
getattr[NULL] /var
   unique: 20, success, outsize: 120
unique: 22, opcode: LOOKUP (1), nodeid: 1, insize: 44, pid: 3837234
LOOKUP /var
getattr[NULL] /var
   NODEID: 2
   unique: 22, success, outsize: 144
   ...
   unique: 278, opcode: GETATTR (3), nodeid: 34, insize: 56, pid: 3837234
getattr[NULL] /scratch
   unique: 278, success, outsize: 120
unique: 280, opcode: READDIRPLUS (44), nodeid: 1, insize: 80, pid: 3837234
readdirplus[140180484788512] from 512
   unique: 280, success, outsize: 16
unique: 282, opcode: RELEASEDIR (29), nodeid: 1, insize: 64, pid: 0
releasedir[140180484788512] flags: 0x0
   unique: 282, success, outsize: 16
```


Unpatched:

```
bschubert2@imesrv6 example>./passthrough_fh -d -oentry_timeout=60 /tmp/mnt/
FUSE library version: 3.16.2
nullpath_ok: 0
unique: 2, opcode: INIT (26), nodeid: 0, insize: 104, pid: 0
INIT: 7.38
flags=0x73fffffb
max_readahead=0x00020000
   INIT: 7.39
   flags=0x4040f439
   max_readahead=0x00020000
   max_write=0x00100000
   max_background=0
   congestion_threshold=0
   time_gran=1
   unique: 2, success, outsize: 80
unique: 4, opcode: GETATTR (3), nodeid: 1, insize: 56, pid: 3838254
getattr[NULL] /
   unique: 4, success, outsize: 120
unique: 6, opcode: GETXATTR (22), nodeid: 1, insize: 65, pid: 3838254
getxattr / security.selinux 255
   unique: 6, error: -61 (No data available), outsize: 16
unique: 8, opcode: GETXATTR (22), nodeid: 1, insize: 72, pid: 3838254
getxattr / system.posix_acl_access 4096
   unique: 8, error: -61 (No data available), outsize: 16
unique: 10, opcode: GETXATTR (22), nodeid: 1, insize: 73, pid: 3838254
getxattr / system.posix_acl_default 4096
   unique: 10, error: -61 (No data available), outsize: 16
unique: 12, opcode: OPENDIR (27), nodeid: 1, insize: 48, pid: 3838254
opendir flags: 0x18800 /
   opendir[140390669750448] flags: 0x18800 /
   unique: 12, success, outsize: 32
unique: 14, opcode: GETATTR (3), nodeid: 1, insize: 56, pid: 3838254
getattr[NULL] /
   unique: 14, success, outsize: 120
unique: 16, opcode: READDIRPLUS (44), nodeid: 1, insize: 80, pid: 3838254
readdirplus[140390669750448] from 0
   unique: 16, success, outsize: 4016
unique: 18, opcode: LOOKUP (1), nodeid: 1, insize: 44, pid: 3838254
LOOKUP /var
getattr[NULL] /var
   NODEID: 2
   unique: 18, success, outsize: 144
unique: 20, opcode: GETATTR (3), nodeid: 2, insize: 56, pid: 3838254
getattr[NULL] /var
   unique: 20, success, outsize: 120
unique: 22, opcode: LOOKUP (1), nodeid: 1, insize: 44, pid: 3838254
LOOKUP /var
getattr[NULL] /var
   NODEID: 2
   unique: 22, success, outsize: 144
unique: 24, opcode: GETXATTR (22), nodeid: 2, insize: 65, pid: 3838254
getxattr /var security.selinux 255
   unique: 24, error: -61 (No data available), outsize: 16
   ....
   unique: 278, opcode: GETATTR (3), nodeid: 34, insize: 56, pid: 3838254
getattr[NULL] /scratch
   unique: 278, success, outsize: 120
unique: 280, opcode: READDIRPLUS (44), nodeid: 1, insize: 80, pid: 3838254
readdirplus[140390669750448] from 512
   unique: 280, success, outsize: 16
unique: 282, opcode: RELEASEDIR (29), nodeid: 1, insize: 64, pid: 0
releasedir[140390669750448] flags: 0x0
   unique: 282, success, outsize: 16
```

How are you testing your patch?

**Comment by bsbernd** [comment]
I need to check on the kernel side why readdir-plus is not working.

**Comment by bsbernd** [comment]
Hmm, ok, that is not going to have readplus-advantage
```
[ 1079.501705] fuse: fuse_direntplus_link:265 Added dentry sbin timout: 0:0
```

**Comment by bsbernd** [comment]
So after seeing that kernel had received entries without any timeout, this patch fixes readdir-plus for passthrough_fh
```
diff --git a/example/passthrough_fh.c b/example/passthrough_fh.c
index 803de0c7..85457494 100644
--- a/example/passthrough_fh.c
+++ b/example/passthrough_fh.c
@@ -54,6 +54,7 @@ static void *xmp_init(struct fuse_conn_info *conn,
        cfg->use_ino = 1;
        cfg->nullpath_ok = 1;

+#if 0
        /* Pick up changes from lower filesystem right away. This is
           also necessary for better hardlink support. When the kernel
           calls the unlink() handler, it does not know the inode of
@@ -64,6 +65,7 @@ static void *xmp_init(struct fuse_conn_info *conn,
        cfg->entry_timeout = 0;
        cfg->attr_timeout = 0;
        cfg->negative_timeout = 0;
+#endif

        return NULL;
 }
 ```
 
 I.e.  I see readdir-plus working without your patch
 ```
 bernd@squeeze1 build-debian>fusermount -u /tmp/mnt/; example/passthrough_fh -d -oentry_timeout=60 /tmp/mnt/
fusermount: entry for /tmp/mnt not found in /etc/mtab
FUSE library version: 3.16.2
nullpath_ok: 0
unique: 2, opcode: INIT (26), nodeid: 0, insize: 104, pid: 0
INIT: 7.39
flags=0x73fffffb
max_readahead=0x00020000
   INIT: 7.39
   flags=0x4040f439
   max_readahead=0x00020000
   max_write=0x00100000
   max_background=0
   congestion_threshold=0
   time_gran=1
   unique: 2, success, outsize: 80
unique: 4, opcode: GETATTR (3), nodeid: 1, insize: 56, pid: 1391
getattr[NULL] /
   unique: 4, success, outsize: 120
unique: 6, opcode: GETXATTR (22), nodeid: 1, insize: 65, pid: 1391
getxattr / security.selinux 255
   unique: 6, error: -95 (Operation not supported), outsize: 16
unique: 8, opcode: GETXATTR (22), nodeid: 1, insize: 72, pid: 1391
getxattr / system.posix_acl_access 4096
   unique: 8, error: -95 (Operation not supported), outsize: 16
unique: 10, opcode: OPENDIR (27), nodeid: 1, insize: 48, pid: 1391
opendir flags: 0x18800 /
   opendir[105759274702960] flags: 0x18800 /
   unique: 10, success, outsize: 32
unique: 12, opcode: READDIRPLUS (44), nodeid: 1, insize: 80, pid: 1391
readdirplus[105759274702960] from 0
offset=0 d->offset=0
sbin fill_flags=2
tmpa fill_flags=2
home fill_flags=2
lustre fill_flags=2
etc fill_flags=2
foo fill_flags=2
usr fill_flags=2
cgroup.procs fill_flags=2
bar fill_flags=2
file1 fill_flags=2
proc fill_flags=2
root fill_flags=2
lib64 fill_flags=2
vmlinuz.old fill_flags=2
run fill_flags=2
lib fill_flags=2
initrd.img fill_flags=2
boot fill_flags=2
sys fill_flags=2
dev fill_flags=2
fhgfs fill_flags=2
media fill_flags=2
scratch fill_flags=2
opt fill_flags=2
file2 fill_flags=2
tmp fill_flags=2
   unique: 12, success, outsize: 4040
unique: 14, opcode: READLINK (5), nodeid: 15, insize: 40, pid: 1391
readlink /vmlinuz.old 4097
   unique: 14, success, outsize: 39
unique: 16, opcode: LOOKUP (1), nodeid: 19, insize: 59, pid: 1391
LOOKUP /boot/vmlinuz-6.5.0-rc6+
getattr[NULL] /boot/vmlinuz-6.5.0-rc6+
   unique: 16, error: -2 (No such file or directory), outsize: 16
unique: 18, opcode: READLINK (5), nodeid: 18, insize: 40, pid: 1391
readlink /initrd.img 4097
   unique: 18, success, outsize: 47
unique: 20, opcode: LOOKUP (1), nodeid: 19, insize: 67, pid: 1391
LOOKUP /boot/initrd.img-5.10.0-22-amd64
getattr[NULL] /boot/initrd.img-5.10.0-22-amd64
   NODEID: 28
   unique: 20, success, outsize: 144
unique: 22, opcode: READDIRPLUS (44), nodeid: 1, insize: 80, pid: 1391
readdirplus[105759274702960] from 7000290212543405504
offset=7000290212543405504 d->offset=7000290212543405504
tmp fill_flags=2
mnt fill_flags=2
srv fill_flags=2
initrd fill_flags=2
lost+found fill_flags=2
initrd.img.old fill_flags=2
s fill_flags=2
chroot fill_flags=2
var fill_flags=2
root.fh fill_flags=2
bin fill_flags=2
vmlinuz fill_flags=2
core fill_flags=2
   unique: 22, success, outsize: 2112
unique: 24, opcode: READLINK (5), nodeid: 33, insize: 40, pid: 1391
readlink /initrd.img.old 4097
   unique: 24, success, outsize: 42
unique: 26, opcode: LOOKUP (1), nodeid: 19, insize: 62, pid: 1391
LOOKUP /boot/initrd.img-6.5.0-rc6+
getattr[NULL] /boot/initrd.img-6.5.0-rc6+
   unique: 26, error: -2 (No such file or directory), outsize: 16
unique: 28, opcode: READLINK (5), nodeid: 39, insize: 40, pid: 1391
readlink /vmlinuz 4097
   unique: 28, success, outsize: 44
unique: 30, opcode: LOOKUP (1), nodeid: 19, insize: 64, pid: 1391
LOOKUP /boot/vmlinuz-5.10.0-22-amd64
getattr[NULL] /boot/vmlinuz-5.10.0-22-amd64
   NODEID: 41
   unique: 30, success, outsize: 144
unique: 32, opcode: READDIRPLUS (44), nodeid: 1, insize: 80, pid: 1391
readdirplus[105759274702960] from 9223372036854775807
offset=9223372036854775807 d->offset=9223372036854775807
   unique: 32, success, outsize: 16
unique: 34, opcode: RELEASEDIR (29), nodeid: 1, insize: 64, pid: 0
releasedir[105759274702960] flags: 0x0
   unique: 34, success, outsize: 16
 ```

**Comment by bsbernd** [comment]
@farlongsignal Thanks for the info, I will check during the next days again what is the actual issue with 0 offsets. When you initially write 0 offsets, I thought you meant the kernel request, but it is actually the offset for the filler.

**Review by bsbernd — APPROVED** [review]
I just took some time to look into this code (and to understand it). Thanks this change makes sense! Although I think we need to add comments to the code and maybe refactor readdir code a bit.

**Pull request merged**